File: app/src/main/java/com/ntl/guidelinesapp/modules/dagger2/di/modules/ViewModelModule.java
package com.ntl.guidelinesapp.modules.dagger2.di.modules;

import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

import com.ntl.guidelinesapp.modules.dagger2.di.ViewModelKey;
import com.ntl.guidelinesapp.modules.dagger2.viewmodel.UserViewModel;
import com.ntl.guidelinesapp.modules.dagger2.viewmodel.ViewModelFactory;

import dagger.Binds;
import dagger.Module;
import dagger.multibindings.IntoMap;

@Module
public abstract class ViewModelModule {
    @Binds
    @IntoMap
    @ViewModelKey(UserViewModel.class)
    abstract ViewModel bindViewModel(UserViewModel userViewModel);

    @Binds
    abstract ViewModelProvider.Factory bindFactory(ViewModelFactory viewModelFactory);
}

File: app/src/main/java/com/ntl/guidelinesapp/modules/kotlin_koin_mvvm/di/ViewModelModule.kt
package com.ntl.guidelinesapp.modules.kotlin_koin_mvvm.di

import com.ntl.guidelinesapp.modules.kotlin_koin_mvvm.viewmodel.MainViewModel
import org.koin.androidx.viewmodel.dsl.viewModel
import org.koin.dsl.module

val ViewModelModule = module {

    viewModel { MainViewModel(get()) }
}